File: src/builtins/function.rs
use crate::*;
use std::rc::Rc;

#[derive(Clone)]
pub struct Function(pub Rc<dyn Fn(&Value, &mut Environment, &Stack) -> Result>);

impl Function {
    pub fn new(function: impl Fn(&Value, &mut Environment, &Stack) -> Result + 'static) -> Self {
        Function(Rc::new(function))
    }
}

primitive!(function for Function);

impl Value {
    pub fn call(&self, parameter: &Value, env: &mut Environment, stack: &Stack) -> Result {
        let stack = stack.add(|| format!("Calling '{}'", self.format(env, stack)));

        let function = self.get_primitive_or::<Function>(
            "Cannot call this value because it does not have the Function trait",
            env,
            &stack,
        )?;

        function.0(parameter, env, &stack)
    }
}

pub(crate) fn setup(env: &mut Environment) {
    // Function ::= Text
    env.add_conformance_for_primitive(TraitID::text(), |_: Function, _, _| {
        Ok(Some(Value::of(Text {
            text: String::from("<function>"),
            location: None,
        })))
    });
}

File: Sources/Builtins/Function.swift
public typealias Function = (Value, inout Environment, ProgramStack) throws -> Value

extension TraitID where T == Function {
    public static let function = TraitID(debugLabel: "Function")
}

extension Trait where T == Function {
    public static func function(_ value: @escaping Function) -> Self {
        Trait(id: .function) { _, _ in value }
    }
}

extension Value {
    public func call(
        with parameter: Value,
        _ env: inout Environment,
        _ stack: ProgramStack
    ) throws -> Value {
        let stack = stack.add("Calling '\(self.format(&env, stack))'")

        let function = try self.trait(
            .function,
            orError: "Cannot call this value because it does not have the Function trait",
            &env,
            stack
        )

        return try function(parameter, &env, stack)
    }
}
